### ISSUE TITLE
Program received and didn't handle signal ILL (4)

### ISSUE THREAD

**Issue by matthewmcgarvey** [crystal-lang/crystal#12831]
## Bug Report

When running code that compiled fine, my app crashed with the message of "Program received and didn't handle signal ILL (4)".

## Reproduce

I'm sorry, I don't have a minimal reproduction as I don't understand how it's happening, but I did make a reproducible example within my SQL query builder library. The commit message explains what you need to do to trigger the error. https://github.com/matthewmcgarvey/focus/commit/10e0ea2a12643688043e2740de92470e71248aca

## Additional Comments

I was able to fix the error in this commit https://github.com/matthewmcgarvey/focus/commit/e8708f09b8e98bcb5d2a48465c65bdad343dad22

I don't understand what I did wrong and how this fixed it. There are 2 or 3 different changes in the commit (besides just applying the changes throughout the codebase), but the one that actually fixed it was adding the second generic type to `BinaryExpression`. Hopefully this can be a good reference for what might be wrong or if this ticket is a duplicate.

I did find https://github.com/crystal-lang/crystal/issues/9667#issuecomment-710405767 and think that it could be related, but I wasn't sure. 

## Specs

Crystal 1.6.2
macOS 12.6

**Comment by straight-shoota** [comment]
Yes, I think it could be related to https://github.com/crystal-lang/crystal/issues/9667#issuecomment-710405767
At least it produces the same behaviour. The ILL signal seems to be only happening on macOS. It does not reproduce on Linux, but there's some odd behaviour there as well.

I'm not sure if this is *actually* the same problem you're facing, but with this simple example it's much easier to dive in. Maybe we can find a fix and if it works for your complex example, we have a win.

This is the code in question (originally posted in https://github.com/crystal-lang/crystal/issues/9667#issuecomment-710405767, I added the ABC indicators).
```cr
abstract class Foo
end

class Bar(T) < Foo
end

puts "A"
p Bar.as(Foo.class).name
puts "B"
p Bar(Int32).as(Foo.class).name
puts "C"
```

On macOS it prints `A` and then immediately aborts with "Program received and didn't handle signal ILL (4)" as it executes the first class name inspection.
On linux it prints only `A\n"Bar(Int32)"\bC` and exits successfully. The middle two parts are completely skipped!

When removing either of the two class name expressions, the program executes fully as expected.
